llvm/lib/Support/SHA1.cpp

```diff
@@ -307,3 +307,36 @@ std::array<uint8_t, 20> SHA1::hash(ArrayRef<uint8_t> Data) {
   memcpy(Arr.data(), S.data(), S.size());
   return Arr;
 }
+
+std::string SHA1::hashToString(ArrayRef<uint8_t> Data, StringRef Mapping) {
+  constexpr auto BitsPerChar = 6;
+  constexpr auto CombinationPerChar = 1 << BitsPerChar;
+  constexpr auto LoopIterations = 27;
+  assert(Mapping.size() >= CombinationPerChar);
+  std::array<uint8_t, 20> Hash = SHA1::hash(Data);
+  std::string Res;
+  Res.reserve(LoopIterations);
+
+  unsigned Bits = 0;
+  unsigned Elem = 0;
+  while (1) {
+    uint16_t Value = Hash[Elem] << 8;
```

Comment:
Too complex...
This should be in the loop prologue, before the loop.
No need to reload every time this.
You can remove a lot of shifts if you load the new bits by the MSB and extract by LSB.
```c++
uint16_t Value = Hash[Elem];
```